{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5895c995",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from CRISPResso2 import CRISPRessoShared, CRISPRessoPlot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c443b274",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10a.FANC.Substitution_frequencies_at_each_bp.pdf\n",
      "10a.FANC.Substitution_frequencies_at_each_bp.png\n",
      "10a.HDR.Substitution_frequencies_at_each_bp.pdf\n",
      "10a.HDR.Substitution_frequencies_at_each_bp.png\n",
      "10b.FANC.Substitution_frequency_barplot.pdf\n",
      "10b.FANC.Substitution_frequency_barplot.png\n",
      "10b.HDR.Substitution_frequency_barplot.pdf\n",
      "10b.HDR.Substitution_frequency_barplot.png\n",
      "10c.FANC.Substitution_frequency_barplot_in_quantification_window.pdf\n",
      "10c.FANC.Substitution_frequency_barplot_in_quantification_window.png\n",
      "10c.HDR.Substitution_frequency_barplot_in_quantification_window.pdf\n",
      "10c.HDR.Substitution_frequency_barplot_in_quantification_window.png\n",
      "10d.FANC.Log2_nucleotide_frequency_around_dear.pdf\n",
      "10d.FANC.Log2_nucleotide_frequency_around_dear.png\n",
      "10d.FANC.Log2_nucleotide_frequency_around_hi.pdf\n",
      "10d.FANC.Log2_nucleotide_frequency_around_hi.png\n",
      "10d.FANC.Log2_nucleotide_frequency_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "10d.FANC.Log2_nucleotide_frequency_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "10d.HDR.Log2_nucleotide_frequency_around_dear.pdf\n",
      "10d.HDR.Log2_nucleotide_frequency_around_dear.png\n",
      "10d.HDR.Log2_nucleotide_frequency_around_hi.pdf\n",
      "10d.HDR.Log2_nucleotide_frequency_around_hi.png\n",
      "10d.HDR.Log2_nucleotide_frequency_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "10d.HDR.Log2_nucleotide_frequency_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "10e.FANC.Selected_conversion_at_Cs_around_dear.pdf\n",
      "10e.FANC.Selected_conversion_at_Cs_around_dear.png\n",
      "10e.FANC.Selected_conversion_at_Cs_around_hi.pdf\n",
      "10e.FANC.Selected_conversion_at_Cs_around_hi.png\n",
      "10e.FANC.Selected_conversion_at_Cs_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "10e.FANC.Selected_conversion_at_Cs_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "10e.HDR.Selected_conversion_at_Cs_around_dear.pdf\n",
      "10e.HDR.Selected_conversion_at_Cs_around_dear.png\n",
      "10e.HDR.Selected_conversion_at_Cs_around_hi.pdf\n",
      "10e.HDR.Selected_conversion_at_Cs_around_hi.png\n",
      "10e.HDR.Selected_conversion_at_Cs_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "10e.HDR.Selected_conversion_at_Cs_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "10f.FANC.Selected_conversion_no_ref_at_Cs_around_dear.pdf\n",
      "10f.FANC.Selected_conversion_no_ref_at_Cs_around_dear.png\n",
      "10f.FANC.Selected_conversion_no_ref_at_Cs_around_hi.pdf\n",
      "10f.FANC.Selected_conversion_no_ref_at_Cs_around_hi.png\n",
      "10f.FANC.Selected_conversion_no_ref_at_Cs_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "10f.FANC.Selected_conversion_no_ref_at_Cs_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "10f.HDR.Selected_conversion_no_ref_at_Cs_around_dear.pdf\n",
      "10f.HDR.Selected_conversion_no_ref_at_Cs_around_dear.png\n",
      "10f.HDR.Selected_conversion_no_ref_at_Cs_around_hi.pdf\n",
      "10f.HDR.Selected_conversion_no_ref_at_Cs_around_hi.png\n",
      "10f.HDR.Selected_conversion_no_ref_at_Cs_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "10f.HDR.Selected_conversion_no_ref_at_Cs_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "10g.FANC.Selected_conversion_no_ref_scaled_at_Cs_around_dear.pdf\n",
      "10g.FANC.Selected_conversion_no_ref_scaled_at_Cs_around_dear.png\n",
      "10g.FANC.Selected_conversion_no_ref_scaled_at_Cs_around_hi.pdf\n",
      "10g.FANC.Selected_conversion_no_ref_scaled_at_Cs_around_hi.png\n",
      "10g.FANC.Selected_conversion_no_ref_scaled_at_Cs_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "10g.FANC.Selected_conversion_no_ref_scaled_at_Cs_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "10g.HDR.Selected_conversion_no_ref_scaled_at_Cs_around_dear.pdf\n",
      "10g.HDR.Selected_conversion_no_ref_scaled_at_Cs_around_dear.png\n",
      "10g.HDR.Selected_conversion_no_ref_scaled_at_Cs_around_hi.pdf\n",
      "10g.HDR.Selected_conversion_no_ref_scaled_at_Cs_around_hi.png\n",
      "10g.HDR.Selected_conversion_no_ref_scaled_at_Cs_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "10g.HDR.Selected_conversion_no_ref_scaled_at_Cs_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "1a.Read_barplot.pdf\n",
      "1a.Read_barplot.png\n",
      "1b.Alignment_pie_chart.pdf\n",
      "1b.Alignment_pie_chart.png\n",
      "1c.Alignment_barplot.pdf\n",
      "1c.Alignment_barplot.png\n",
      "1d.Detection_of_dsODN.pdf\n",
      "1d.Detection_of_dsODN.png\n",
      "2a.FANC.Nucleotide_percentage_quilt.pdf\n",
      "2a.FANC.Nucleotide_percentage_quilt.png\n",
      "2a.HDR.Nucleotide_percentage_quilt.pdf\n",
      "2a.HDR.Nucleotide_percentage_quilt.png\n",
      "2b.FANC.Nucleotide_percentage_quilt_around_dear.pdf\n",
      "2b.FANC.Nucleotide_percentage_quilt_around_dear.png\n",
      "2b.FANC.Nucleotide_percentage_quilt_around_hi.pdf\n",
      "2b.FANC.Nucleotide_percentage_quilt_around_hi.png\n",
      "2b.FANC.Nucleotide_percentage_quilt_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "2b.FANC.Nucleotide_percentage_quilt_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "2b.HDR.Nucleotide_percentage_quilt_around_dear.pdf\n",
      "2b.HDR.Nucleotide_percentage_quilt_around_dear.png\n",
      "2b.HDR.Nucleotide_percentage_quilt_around_hi.pdf\n",
      "2b.HDR.Nucleotide_percentage_quilt_around_hi.png\n",
      "2b.HDR.Nucleotide_percentage_quilt_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "2b.HDR.Nucleotide_percentage_quilt_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "3a.FANC.Indel_size_distribution.pdf\n",
      "3a.FANC.Indel_size_distribution.png\n",
      "3a.HDR.Indel_size_distribution.pdf\n",
      "3a.HDR.Indel_size_distribution.png\n",
      "3b.FANC.Insertion_deletion_substitutions_size_hist.pdf\n",
      "3b.FANC.Insertion_deletion_substitutions_size_hist.png\n",
      "3b.HDR.Insertion_deletion_substitutions_size_hist.pdf\n",
      "3b.HDR.Insertion_deletion_substitutions_size_hist.png\n",
      "4a.FANC.Combined_insertion_deletion_substitution_locations.pdf\n",
      "4a.FANC.Combined_insertion_deletion_substitution_locations.png\n",
      "4a.HDR.Combined_insertion_deletion_substitution_locations.pdf\n",
      "4a.HDR.Combined_insertion_deletion_substitution_locations.png\n",
      "4b.FANC.Insertion_deletion_substitution_locations.pdf\n",
      "4b.FANC.Insertion_deletion_substitution_locations.png\n",
      "4b.HDR.Insertion_deletion_substitution_locations.pdf\n",
      "4b.HDR.Insertion_deletion_substitution_locations.png\n",
      "4c.FANC.Quantification_window_insertion_deletion_substitution_locations.pdf\n",
      "4c.FANC.Quantification_window_insertion_deletion_substitution_locations.png\n",
      "4c.HDR.Quantification_window_insertion_deletion_substitution_locations.pdf\n",
      "4c.HDR.Quantification_window_insertion_deletion_substitution_locations.png\n",
      "4d.FANC.Position_dependent_average_indel_size.pdf\n",
      "4d.FANC.Position_dependent_average_indel_size.png\n",
      "4d.HDR.Position_dependent_average_indel_size.pdf\n",
      "4d.HDR.Position_dependent_average_indel_size.png\n",
      "4e.FANC.Global_mutations_in_all_reads.pdf\n",
      "4e.FANC.Global_mutations_in_all_reads.png\n",
      "4f.FANC.Global_mutations_in_HDR_reads_with_reference_to_FANC.pdf\n",
      "4f.FANC.Global_mutations_in_HDR_reads_with_reference_to_FANC.png\n",
      "4g.HDR_nucleotide_percentage_quilt.pdf\n",
      "4g.HDR_nucleotide_percentage_quilt.png\n",
      "5.HDR.Frameshift_in-frame_mutations_pie_chart.pdf\n",
      "5.HDR.Frameshift_in-frame_mutations_pie_chart.png\n",
      "5a.Global_frameshift_in-frame_mutations_pie_chart.pdf\n",
      "5a.Global_frameshift_in-frame_mutations_pie_chart.png\n",
      "6.FANC.Frameshift_in-frame_mutation_profiles.pdf\n",
      "6.FANC.Frameshift_in-frame_mutation_profiles.png\n",
      "6.HDR.Frameshift_in-frame_mutation_profiles.pdf\n",
      "6.HDR.Frameshift_in-frame_mutation_profiles.png\n",
      "6a.Global_frameshift_in-frame_mutation_profiles.pdf\n",
      "6a.Global_frameshift_in-frame_mutation_profiles.png\n",
      "7.FANC.Insertion_deletion_substitution_locations_noncoding.pdf\n",
      "7.FANC.Insertion_deletion_substitution_locations_noncoding.png\n",
      "7.HDR.Insertion_deletion_substitution_locations_noncoding.pdf\n",
      "7.HDR.Insertion_deletion_substitution_locations_noncoding.png\n",
      "8.FANC.Potential_splice_sites_pie_chart.pdf\n",
      "8.FANC.Potential_splice_sites_pie_chart.png\n",
      "8.HDR.Potential_splice_sites_pie_chart.pdf\n",
      "8.HDR.Potential_splice_sites_pie_chart.png\n",
      "8a.Global_potential_splice_sites_pie_chart.pdf\n",
      "8a.Global_potential_splice_sites_pie_chart.png\n",
      "9.FANC.Alleles_frequency_table_around_dear.pdf\n",
      "9.FANC.Alleles_frequency_table_around_dear.png\n",
      "9.FANC.Alleles_frequency_table_around_hi.pdf\n",
      "9.FANC.Alleles_frequency_table_around_hi.png\n",
      "9.FANC.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "9.FANC.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "9.HDR.Alleles_frequency_table_around_dear.pdf\n",
      "9.HDR.Alleles_frequency_table_around_dear.png\n",
      "9.HDR.Alleles_frequency_table_around_hi.pdf\n",
      "9.HDR.Alleles_frequency_table_around_hi.png\n",
      "9.HDR.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.pdf\n",
      "9.HDR.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.png\n",
      "Alleles_frequency_table.zip\n",
      "CRISPResso2_info.json\n",
      "CRISPResso2_report.html\n",
      "CRISPResso_RUNNING_LOG.txt\n",
      "CRISPResso_mapping_statistics.txt\n",
      "CRISPResso_quantification_of_editing_frequency.txt\n",
      "CRISPResso_reference_info.txt\n",
      "CRISPResso_status.json\n",
      "FANC.Alleles_frequency_table_around_dear.txt\n",
      "FANC.Alleles_frequency_table_around_hi.txt\n",
      "FANC.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt\n",
      "FANC.Cas9_filtered.fastq.gz\n",
      "FANC.Cut_points.json\n",
      "FANC.Deletion_histogram.txt\n",
      "FANC.Effect_vector_combined.npz\n",
      "FANC.Effect_vector_combined.txt\n",
      "FANC.Effect_vector_deletion.npz\n",
      "FANC.Effect_vector_deletion.txt\n",
      "FANC.Effect_vector_deletion_noncoding.txt\n",
      "FANC.Effect_vector_insertion.npz\n",
      "FANC.Effect_vector_insertion.txt\n",
      "FANC.Effect_vector_insertion_noncoding.txt\n",
      "FANC.Effect_vector_substitution.npz\n",
      "FANC.Effect_vector_substitution.txt\n",
      "FANC.Effect_vector_substitution_noncoding.txt\n",
      "FANC.Frameshift_analysis.txt\n",
      "FANC.Indel_histogram.txt\n",
      "FANC.Insertion_histogram.txt\n",
      "FANC.Modification_count_vectors.txt\n",
      "FANC.Nucleotide_frequency_table.txt\n",
      "FANC.Nucleotide_percentage_table.txt\n",
      "FANC.Position_dependent_vector_avg_deletion_size.npz\n",
      "FANC.Position_dependent_vector_avg_insertion_size.npz\n",
      "FANC.Quantification_window_modification_count_vectors.txt\n",
      "FANC.Quantification_window_nucleotide_frequency_table.txt\n",
      "FANC.Quantification_window_nucleotide_percentage_table.txt\n",
      "FANC.Quantification_window_substitution_frequency_table.txt\n",
      "FANC.Reads_from_all_amplicons_modification_percent_table.txt\n",
      "FANC.Reads_from_all_amplicons_nucleotide_percent_table.txt\n",
      "FANC.Selected_nucleotide_frequency_table_around_dear.txt\n",
      "FANC.Selected_nucleotide_frequency_table_around_hi.txt\n",
      "FANC.Selected_nucleotide_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt\n",
      "FANC.Selected_nucleotide_percentage_table_around_dear.txt\n",
      "FANC.Selected_nucleotide_percentage_table_around_hi.txt\n",
      "FANC.Selected_nucleotide_percentage_table_around_sgRNA_CCCACTGAAGGCCC.txt\n",
      "FANC.Splice_sites_analysis.txt\n",
      "FANC.Substitution_frequency_table.txt\n",
      "FANC.Substitution_histogram.txt\n",
      "FANC.sgRNA_intervals.json\n",
      "HDR.Alleles_frequency_table_around_dear.txt\n",
      "HDR.Alleles_frequency_table_around_hi.txt\n",
      "HDR.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt\n",
      "HDR.Cut_points.json\n",
      "HDR.Deletion_histogram.txt\n",
      "HDR.Effect_vector_combined.npz\n",
      "HDR.Effect_vector_combined.txt\n",
      "HDR.Effect_vector_deletion.npz\n",
      "HDR.Effect_vector_deletion.txt\n",
      "HDR.Effect_vector_deletion_noncoding.txt\n",
      "HDR.Effect_vector_insertion.npz\n",
      "HDR.Effect_vector_insertion.txt\n",
      "HDR.Effect_vector_insertion_noncoding.txt\n",
      "HDR.Effect_vector_substitution.npz\n",
      "HDR.Effect_vector_substitution.txt\n",
      "HDR.Effect_vector_substitution_noncoding.txt\n",
      "HDR.Frameshift_analysis.txt\n",
      "HDR.Indel_histogram.txt\n",
      "HDR.Insertion_histogram.txt\n",
      "HDR.Modification_count_vectors.txt\n",
      "HDR.Nucleotide_frequency_table.txt\n",
      "HDR.Nucleotide_percentage_table.txt\n",
      "HDR.Position_dependent_vector_avg_deletion_size.npz\n",
      "HDR.Position_dependent_vector_avg_insertion_size.npz\n",
      "HDR.Quantification_window_modification_count_vectors.txt\n",
      "HDR.Quantification_window_nucleotide_frequency_table.txt\n",
      "HDR.Quantification_window_nucleotide_percentage_table.txt\n",
      "HDR.Quantification_window_substitution_frequency_table.txt\n",
      "HDR.Selected_nucleotide_frequency_table_around_dear.txt\n",
      "HDR.Selected_nucleotide_frequency_table_around_hi.txt\n",
      "HDR.Selected_nucleotide_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt\n",
      "HDR.Selected_nucleotide_percentage_table_around_dear.txt\n",
      "HDR.Selected_nucleotide_percentage_table_around_hi.txt\n",
      "HDR.Selected_nucleotide_percentage_table_around_sgRNA_CCCACTGAAGGCCC.txt\n",
      "HDR.Splice_sites_analysis.txt\n",
      "HDR.Substitution_frequency_table.txt\n",
      "HDR.Substitution_histogram.txt\n",
      "HDR.sgRNA_intervals.json\n"
     ]
    }
   ],
   "source": [
    "!ls ../CRISPResso2_tests/cli_integration_tests/CRISPResso_on_params/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a64da86e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#! unzip ../CRISPResso2_tests/cli_integration_tests/CRISPResso_on_params/Alleles_frequency_table.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ebbb2808",
   "metadata": {},
   "outputs": [],
   "source": [
    "# I think this is right\n",
    "df_alleles = pd.read_csv('Alleles_frequency_table.txt', sep='\\t', header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "51eab164",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Aligned_Sequence', 'Reference_Sequence', 'Reference_Name',\n",
       "       'Read_Status', 'n_deleted', 'n_inserted', 'n_mutated', '#Reads',\n",
       "       '%Reads', 'contains dsODN', 'contains dsODN fragment'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_alleles.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "156ae9b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['FANC', 'HDR']\n"
     ]
    }
   ],
   "source": [
    "root = \"../CRISPResso_on_Linear_chr11_sm_201_left\"\n",
    "root = \"../CRISPResso2_tests/cli_integration_tests/CRISPResso_on_params/\"\n",
    "crispresso2_info = CRISPRessoShared.load_crispresso_info(root)\n",
    "print(crispresso2_info['results']['ref_names'])\n",
    "ref_name = crispresso2_info['results']['ref_names'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "22b83821",
   "metadata": {},
   "outputs": [],
   "source": [
    "refs = crispresso2_info['results']['refs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5cfd3b1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'FANC': {'name': 'FANC',\n",
       "  'sequence': 'CGGATGTTCCAATCAGTACGCAGAGAGTCGCCGTCTCCAAGGTGAAAGCGGAAGTAGGGCCTTCGCGCACCTCATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAGCTTCTGGCGGTCTCAAGCACTACCTACGTCAGCACCTGGGACCCCGCCACCGTGCGCCGGGCCTTGCAGTGGGCGCGCTACCTGCGCCACATCCATCGGCGCTTTGGTCGG',\n",
       "  'sequence_length': 223,\n",
       "  'min_aln_score': 80,\n",
       "  'gap_incentive': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0]),\n",
       "  'sgRNA_cut_points': [91, 188, 172],\n",
       "  'sgRNA_intervals': [[75, 94], [171, 191], [170, 184]],\n",
       "  'sgRNA_sequences': ['GGAATCCCTTCTGCAGCACC',\n",
       "   'GGCCTTGCAGTGGGCGCGCTA',\n",
       "   'CCCACTGCAAGGCCC'],\n",
       "  'sgRNA_plot_cut_points': [False, False, False],\n",
       "  'sgRNA_plot_idxs': [[72,\n",
       "    73,\n",
       "    74,\n",
       "    75,\n",
       "    76,\n",
       "    77,\n",
       "    78,\n",
       "    79,\n",
       "    80,\n",
       "    81,\n",
       "    82,\n",
       "    83,\n",
       "    84,\n",
       "    85,\n",
       "    86,\n",
       "    87,\n",
       "    88,\n",
       "    89,\n",
       "    90,\n",
       "    91,\n",
       "    92,\n",
       "    93,\n",
       "    94,\n",
       "    95,\n",
       "    96,\n",
       "    97,\n",
       "    98,\n",
       "    99,\n",
       "    100,\n",
       "    101,\n",
       "    102,\n",
       "    103,\n",
       "    104,\n",
       "    105,\n",
       "    106,\n",
       "    107,\n",
       "    108,\n",
       "    109,\n",
       "    110,\n",
       "    111],\n",
       "   [169,\n",
       "    170,\n",
       "    171,\n",
       "    172,\n",
       "    173,\n",
       "    174,\n",
       "    175,\n",
       "    176,\n",
       "    177,\n",
       "    178,\n",
       "    179,\n",
       "    180,\n",
       "    181,\n",
       "    182,\n",
       "    183,\n",
       "    184,\n",
       "    185,\n",
       "    186,\n",
       "    187,\n",
       "    188,\n",
       "    189,\n",
       "    190,\n",
       "    191,\n",
       "    192,\n",
       "    193,\n",
       "    194,\n",
       "    195,\n",
       "    196,\n",
       "    197,\n",
       "    198,\n",
       "    199,\n",
       "    200,\n",
       "    201,\n",
       "    202,\n",
       "    203,\n",
       "    204,\n",
       "    205,\n",
       "    206,\n",
       "    207,\n",
       "    208],\n",
       "   [153,\n",
       "    154,\n",
       "    155,\n",
       "    156,\n",
       "    157,\n",
       "    158,\n",
       "    159,\n",
       "    160,\n",
       "    161,\n",
       "    162,\n",
       "    163,\n",
       "    164,\n",
       "    165,\n",
       "    166,\n",
       "    167,\n",
       "    168,\n",
       "    169,\n",
       "    170,\n",
       "    171,\n",
       "    172,\n",
       "    173,\n",
       "    174,\n",
       "    175,\n",
       "    176,\n",
       "    177,\n",
       "    178,\n",
       "    179,\n",
       "    180,\n",
       "    181,\n",
       "    182,\n",
       "    183,\n",
       "    184,\n",
       "    185,\n",
       "    186,\n",
       "    187,\n",
       "    188,\n",
       "    189,\n",
       "    190,\n",
       "    191,\n",
       "    192]],\n",
       "  'sgRNA_names': ['hi', 'dear', ''],\n",
       "  'sgRNA_mismatches': [[], [0], [7]],\n",
       "  'sgRNA_orig_sequences': ['GGAATCCCTTCTGCAGCACC',\n",
       "   'AGCCTTGCAGTGGGCGCGCTA',\n",
       "   'CCCACTGAAGGCCC'],\n",
       "  'sgRNA_include_idxs': [[], [], []],\n",
       "  'contains_guide': True,\n",
       "  'contains_coding_seq': True,\n",
       "  'exon_positions': [56,\n",
       "   57,\n",
       "   58,\n",
       "   59,\n",
       "   60,\n",
       "   61,\n",
       "   62,\n",
       "   63,\n",
       "   64,\n",
       "   65,\n",
       "   66,\n",
       "   67,\n",
       "   68,\n",
       "   69,\n",
       "   70,\n",
       "   71,\n",
       "   72,\n",
       "   73,\n",
       "   74,\n",
       "   75,\n",
       "   76,\n",
       "   77,\n",
       "   78,\n",
       "   79,\n",
       "   80,\n",
       "   81,\n",
       "   82,\n",
       "   83,\n",
       "   84,\n",
       "   85,\n",
       "   86,\n",
       "   87,\n",
       "   88,\n",
       "   89,\n",
       "   90,\n",
       "   91,\n",
       "   92,\n",
       "   93,\n",
       "   94,\n",
       "   95,\n",
       "   96,\n",
       "   97,\n",
       "   98,\n",
       "   99,\n",
       "   100,\n",
       "   101,\n",
       "   102,\n",
       "   103,\n",
       "   104,\n",
       "   105,\n",
       "   106],\n",
       "  'exon_len_mods': [0],\n",
       "  'exon_intervals': [[56, 107]],\n",
       "  'splicing_positions': {54, 55, 107, 108},\n",
       "  'include_idxs': array([20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 45, 46, 47, 48, 49, 50]),\n",
       "  'exclude_idxs': array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,\n",
       "          13,  14, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218,\n",
       "         219, 220, 221, 222]),\n",
       "  'idx_cloned_from': None,\n",
       "  'fw_seeds': ['GTACGCAGAG',\n",
       "   'CAGAGAGTCG',\n",
       "   'AGTCGCCGTC',\n",
       "   'CCGTCTCCAA',\n",
       "   'TCCAAGGTGA',\n",
       "   'GGTGAAAGCG',\n",
       "   'AAGCGGAAGT',\n",
       "   'GAAGTAGGGC',\n",
       "   'AGGGCCTTCG',\n",
       "   'CTTCGCGCAC',\n",
       "   'CGCACCTCAT',\n",
       "   'CTCATGGAAT',\n",
       "   'GGAATCCCTT',\n",
       "   'CCCTTCTGCA',\n",
       "   'CTGCAGCACC',\n",
       "   'GCACCTGGAT',\n",
       "   'TGGATCGCTT',\n",
       "   'CGCTTTTCCG',\n",
       "   'TTCCGAGCTT',\n",
       "   'AGCTTCTGGC',\n",
       "   'CTGGCGGTCT',\n",
       "   'GGTCTCAAGC',\n",
       "   'CAAGCACTAC',\n",
       "   'ACTACCTACG',\n",
       "   'CTACGTCAGC',\n",
       "   'TCAGCACCTG',\n",
       "   'ACCTGGGACC',\n",
       "   'GGACCCCGCC',\n",
       "   'CCGCCACCGT',\n",
       "   'ACCGTGCGCC',\n",
       "   'GCGCCGGGCC',\n",
       "   'GGGCCTTGCA',\n",
       "   'TTGCAGTGGG',\n",
       "   'GTGGGCGCGC',\n",
       "   'CGCGCTACCT',\n",
       "   'TACCTGCGCC',\n",
       "   'GCGCCACATC'],\n",
       "  'rc_seeds': ['CTCTGCGTAC',\n",
       "   'CGACTCTCTG',\n",
       "   'GACGGCGACT',\n",
       "   'TTGGAGACGG',\n",
       "   'TCACCTTGGA',\n",
       "   'CGCTTTCACC',\n",
       "   'ACTTCCGCTT',\n",
       "   'GCCCTACTTC',\n",
       "   'CGAAGGCCCT',\n",
       "   'GTGCGCGAAG',\n",
       "   'ATGAGGTGCG',\n",
       "   'ATTCCATGAG',\n",
       "   'AAGGGATTCC',\n",
       "   'TGCAGAAGGG',\n",
       "   'GGTGCTGCAG',\n",
       "   'ATCCAGGTGC',\n",
       "   'AAGCGATCCA',\n",
       "   'CGGAAAAGCG',\n",
       "   'AAGCTCGGAA',\n",
       "   'GCCAGAAGCT',\n",
       "   'AGACCGCCAG',\n",
       "   'GCTTGAGACC',\n",
       "   'GTAGTGCTTG',\n",
       "   'CGTAGGTAGT',\n",
       "   'GCTGACGTAG',\n",
       "   'CAGGTGCTGA',\n",
       "   'GGTCCCAGGT',\n",
       "   'GGCGGGGTCC',\n",
       "   'ACGGTGGCGG',\n",
       "   'GGCGCACGGT',\n",
       "   'GGCCCGGCGC',\n",
       "   'TGCAAGGCCC',\n",
       "   'CCCACTGCAA',\n",
       "   'GCGCGCCCAC',\n",
       "   'AGGTAGCGCG',\n",
       "   'GGCGCAGGTA',\n",
       "   'GATGTGGCGC'],\n",
       "  'aln_genome': None,\n",
       "  'aln_chr': None,\n",
       "  'aln_start': None,\n",
       "  'aln_end': None,\n",
       "  'aln_strand': None,\n",
       "  'min_cut': 91,\n",
       "  'max_cut': 188,\n",
       "  'y_values_mut': array([192,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,\n",
       "           0,   0,   0]),\n",
       "  'x_bins_mut': array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),\n",
       "  'y_values_ins': array([192,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,\n",
       "           0,   0,   0]),\n",
       "  'x_bins_ins': array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),\n",
       "  'y_values_del': array([192,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,\n",
       "           0,   0,   0]),\n",
       "  'x_bins_del': array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),\n",
       "  'hdensity': array([  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,\n",
       "           0,   0, 192,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,\n",
       "           0,   0,   0,   0,   0]),\n",
       "  'hlengths': array([-15, -14, -13, -12, -11, -10,  -9,  -8,  -7,  -6,  -5,  -4,  -3,\n",
       "          -2,  -1,   0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,\n",
       "          11,  12,  13,  14,  15]),\n",
       "  'center_index': 15,\n",
       "  'ref_plot_name': 'FANC.',\n",
       "  'insertion_pct_vector_filename': 'FANC.Effect_vector_insertion.txt',\n",
       "  'deletion_pct_vector_filename': 'FANC.Effect_vector_deletion.txt',\n",
       "  'substitution_pct_vector_filename': 'FANC.Effect_vector_substitution.txt',\n",
       "  'combined_pct_vector_filename': 'FANC.Effect_vector_combined.txt',\n",
       "  'quant_window_mod_count_filename': 'FANC.Quantification_window_modification_count_vectors.txt',\n",
       "  'mod_count_filename': 'FANC.Modification_count_vectors.txt',\n",
       "  'mod_count_filename_caption': 'A tab-separated file showing the number of modifications for each position in the amplicon. The first row shows the amplicon sequence, and successive rows show the number of reads with insertions (row 2), insertions_left (row 3), deletions (row 4), substitutions (row 5) and the sum of all modifications (row 6).Additionally, the last row shows the number of reads aligned. If an insertion occurs between bases 5 and 6, the insertions vector will be incremented at bases 5 and 6. However, the insertions_left vector will only be incremented at base 5 so the sum of the insertions_left row represents an accurate count of the number of insertions, whereas the sum of the insertions row will yield twice the number of insertions.',\n",
       "  'frameshift_analysis_filename': 'FANC.Frameshift_analysis.txt',\n",
       "  'frameshift_analysis_filename_caption': 'A text file describing the number of noncoding, in-frame, and frameshift mutations. This report file is produced when the amplicon contains a coding sequence.',\n",
       "  'splice_sites_analysis_filename': 'FANC.Splice_sites_analysis.txt',\n",
       "  'splice_sites_analysis_filename_caption': 'A text file describing the number of splicing sites that are unmodified and modified. This file report is produced when the amplicon contains a coding sequence.',\n",
       "  'insertion_pct_vector_noncoding_filename': 'FANC.Effect_vector_insertion_noncoding.txt',\n",
       "  'insertion_pct_vector_noncoding_filename_caption': 'A tab-separated text file with a one-row header that shows the percentage of reads with a noncoding insertion at each base in the FANC sequence. The first column shows the 1-based position of the amplicon, and the second column shows the percentage of reads with a noncoding insertion at that location. This report file is produced when the amplicon contains a coding sequence.',\n",
       "  'deletion_pct_vector_noncoding_filename': 'FANC.Effect_vector_deletion_noncoding.txt',\n",
       "  'deletion_pct_vector_noncoding_filename_caption': 'A tab-separated text file with a one-row header that shows the percentage of reads with a noncoding deletion at each base in the FANC sequence. The first column shows the 1-based position of the amplicon, and the second column shows the percentage of reads with a noncoding deletion at that location. This report file is produced when the amplicon contains a coding sequence.',\n",
       "  'substitution_pct_vector_noncoding_filename': 'FANC.Effect_vector_substitution_noncoding.txt',\n",
       "  'substitution_pct_vector_noncoding_filename_caption': 'A tab-separated text file with a one-row header that shows the percentage of reads with a noncoding substitution at each base in the FANC sequence. The first column shows the 1-based position of the amplicon, and the second column shows the percentage of reads with a nondcoding substitution at that location. This report file is produced when the amplicon contains a coding sequence.',\n",
       "  'indel_histogram_filename': 'FANC.Indel_histogram.txt',\n",
       "  'indel_histogram_filename_caption': 'A tab-separated text file that shows a histogram of the length of indels (both insertions and deletions) in the FANC sequence in the quantification window. Indels outside of the quantification window are not included. The indel_size column shows the number of substitutions, and the fq column shows the number of reads having an indel of that length.',\n",
       "  'insertion_histogram_filename': 'FANC.Insertion_histogram.txt',\n",
       "  'insertion_histogram_filename_caption': 'A tab-separated text file that shows a histogram of the number of insertions in the FANC sequence in the quantification window. Insertions outside of the quantification window are not included. The ins_size column shows the number of insertions, and the fq column shows the number of reads having that number of insertions.',\n",
       "  'deletion_histogram_filename': 'FANC.Deletion_histogram.txt',\n",
       "  'deletion_histogram_filename_caption': 'A tab-separated text file that shows a histogram of the number of deletions in the FANC sequence in the quantification window. Deletions outside of the quantification window are not included. The del_size column shows the number of deletions, and the fq column shows the number of reads having that number of deletions.',\n",
       "  'substitution_histogram_filename': 'FANC.Substitution_histogram.txt',\n",
       "  'substitution_histogram_filename_caption': 'A tab-separated text file that shows a histogram of the number of substitutions in the FANC sequence in the quantification window. Substitutions outside of the quantification window are not included. The sub_size column shows the number of substitutions, and the fq column shows the number of reads having that number of substitutions.',\n",
       "  'quant_window_nuc_freq_filename': 'FANC.Quantification_window_nucleotide_frequency_table.txt',\n",
       "  'quant_window_nuc_pct_filename': 'FANC.Quantification_window_nucleotide_percentage_table.txt',\n",
       "  'nuc_freq_filename': 'FANC.Nucleotide_frequency_table.txt',\n",
       "  'nuc_pct_filename': 'FANC.Nucleotide_percentage_table.txt',\n",
       "  'quant_window_sub_freq_filename': 'FANC.Quantification_window_substitution_frequency_table.txt',\n",
       "  'sub_freq_table_filename': 'FANC.Substitution_frequency_table.txt',\n",
       "  'plot_2a_root': '2a.FANC.Nucleotide_percentage_quilt',\n",
       "  'plot_2a_caption': 'Figure 2a: Nucleotide distribution across amplicon. At each base in the reference amplicon, the percentage of each base as observed in sequencing reads is shown (A = green; C = orange; G = yellow; T = purple). Black bars show the percentage of reads for which that base was deleted. Brown bars between bases show the percentage of reads having an insertion at that position.',\n",
       "  'plot_2a_data': [['Nucleotide frequency table',\n",
       "    'FANC.Nucleotide_frequency_table.txt']],\n",
       "  'plot_2b_roots': ['2b.FANC.Nucleotide_percentage_quilt_around_hi',\n",
       "   '2b.FANC.Nucleotide_percentage_quilt_around_dear',\n",
       "   '2b.FANC.Nucleotide_percentage_quilt_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_2b_captions': ['Figure 2b: Nucleotide distribution around the hi (GGAATCCCTTCTGCAGCACC).',\n",
       "   'Figure 2b: Nucleotide distribution around the dear (AGCCTTGCAGTGGGCGCGCTA).',\n",
       "   'Figure 2b: Nucleotide distribution around the sgRNA CCCACTGAAGGCCC.'],\n",
       "  'plot_2b_datas': [[['Nucleotide frequency in quantification window',\n",
       "     'FANC.Quantification_window_nucleotide_frequency_table.txt']],\n",
       "   [['Nucleotide frequency in quantification window',\n",
       "     'FANC.Quantification_window_nucleotide_frequency_table.txt']],\n",
       "   [['Nucleotide frequency in quantification window',\n",
       "     'FANC.Quantification_window_nucleotide_frequency_table.txt']]],\n",
       "  'plot_3a_root': '3a.FANC.Indel_size_distribution',\n",
       "  'plot_3a_caption': 'Figure 3a: Frequency distribution of alleles with indels (blue) and without indels (red).',\n",
       "  'plot_3a_data': [['Indel histogram', 'FANC.Indel_histogram.txt']],\n",
       "  'plot_3b_root': '3b.FANC.Insertion_deletion_substitutions_size_hist',\n",
       "  'plot_3b_caption': 'Figure 3b: Left panel, frequency distribution of sequence modifications that increase read length with respect to the reference amplicon, classified as insertions (positive indel size). Middle panel, frequency distribution of sequence modifications that reduce read length with respect to the reference amplicon, classified as deletions (negative indel size). Right panel, frequency distribution of sequence modifications that do not alter read length with respect to the reference amplicon, which are classified as substitutions (number of substituted positions shown).',\n",
       "  'plot_3b_data': [['Insertions frequency', 'FANC.Insertion_histogram.txt'],\n",
       "   ['Deletions Frequency', 'FANC.Deletion_histogram.txt'],\n",
       "   ['Substitutions Frequency', 'FANC.Substitution_histogram.txt']],\n",
       "  'plot_4a_root': '4a.FANC.Combined_insertion_deletion_substitution_locations',\n",
       "  'plot_4a_caption': 'Figure 4a: Combined frequency of any modification across the amplicon. Modifications outside of the quantification window are also shown.',\n",
       "  'plot_4a_data': [],\n",
       "  'plot_4b_root': '4b.FANC.Insertion_deletion_substitution_locations',\n",
       "  'plot_4b_caption': 'Figure 4b: Frequency of insertions, deletions, and substitutions across the entire amplicon, including modifications outside of the quantification window.',\n",
       "  'plot_4b_data': [['Modification frequency',\n",
       "    'HDR.Modification_count_vectors.txt']],\n",
       "  'plot_4c_root': '4c.FANC.Quantification_window_insertion_deletion_substitution_locations',\n",
       "  'plot_4c_caption': 'Figure 4c: Frequency of insertions, deletions, and substitutions across the entire amplicon, considering only modifications that overlap with the quantification window.',\n",
       "  'plot_4c_data': [['Modification frequency in quantification window',\n",
       "    'HDR.Quantification_window_modification_count_vectors.txt']],\n",
       "  'plot_4d_root': '4d.FANC.Position_dependent_average_indel_size',\n",
       "  'plot_4d_caption': 'Figure 4d: Position dependent insertion size(left) and deletion size (right), including only modifications that overlap with the quantification window.',\n",
       "  'plot_4d_data': [],\n",
       "  'plot_4e_root': '4e.FANC.Global_mutations_in_all_reads',\n",
       "  'plot_4e_caption': 'Figure 4e: Positions of modifications in all reads when aligned to the reference sequence (FANC). Insertions: red, deletions: purple, substitutions: green. All modifications (including those outside the quantification window) are shown.',\n",
       "  'plot_4e_data': [],\n",
       "  'plot_4g_root': '4g.HDR_nucleotide_percentage_quilt',\n",
       "  'plot_4g_caption': 'Figure 4g: Nucleotide distribution across all amplicons. At each base in the reference amplicon, the percentage of each base as observed in sequencing reads is shown (A = green; C = orange; G = yellow; T = purple). Black bars show the percentage of reads for which that base was deleted. Brown bars between bases show the percentage of reads having an insertion at that position.',\n",
       "  'plot_4g_data': [['Nucleotide percentage table for all reads aligned to FANC',\n",
       "    'FANC.Reads_from_all_amplicons_nucleotide_percent_table.txt'],\n",
       "   ['Modification percentage table for all reads aligned to FANC',\n",
       "    'FANC.Reads_from_all_amplicons_modification_percent_table.txt'],\n",
       "   ['Nucleotide frequency table for FANC',\n",
       "    'FANC.Nucleotide_frequency_table.txt']],\n",
       "  'plot_6_root': '6.FANC.Frameshift_in-frame_mutation_profiles',\n",
       "  'plot_6_caption': 'Figure 6: Frameshift and in-frame mutagenesis profiles indicating position affected by modification. The y axis shows the number of reads and percentage of all reads in that category (frameshifted (top) or in-frame (bottom)). 0 reads with no length modifications are not shown.',\n",
       "  'plot_6_data': [['Indel histogram for FANC', 'FANC.Indel_histogram.txt']],\n",
       "  'plot_7_root': '7.FANC.Insertion_deletion_substitution_locations_noncoding',\n",
       "  'plot_7_caption': 'Figure 7: Reads with insertions, deletions, and substitutions mapped to reference amplicon position exclusively in noncoding region/s (that is, without mutations affecting coding sequences). The predicted cleavage site is indicated by a vertical dashed line. Only sequence positions directly adjacent to insertions or directly affected by deletions or substitutions are plotted.',\n",
       "  'plot_7_data': [],\n",
       "  'plot_8_root': '8.FANC.Potential_splice_sites_pie_chart',\n",
       "  'plot_8_caption': 'Figure 8: Predicted impact on splice sites. Potential splice sites modified refers to reads in which the either of the two intronic positions adjacent to exon junctions are disrupted.',\n",
       "  'plot_8_data': [],\n",
       "  'plot_10a_root': '10a.FANC.Substitution_frequencies_at_each_bp',\n",
       "  'plot_10a_caption': 'Figure 10a: Substitution frequencies across the amplicon.',\n",
       "  'plot_10a_data': [['Nucleotide frequencies',\n",
       "    'FANC.Nucleotide_frequency_table.txt']],\n",
       "  'plot_10b_root': '10b.FANC.Substitution_frequency_barplot',\n",
       "  'plot_10b_caption': 'Figure 10b: Substitution frequencies across the amplicon.',\n",
       "  'plot_10b_data': [['Nucleotide frequencies',\n",
       "    'FANC.Nucleotide_frequency_table.txt']],\n",
       "  'plot_10c_root': '10c.FANC.Substitution_frequency_barplot_in_quantification_window',\n",
       "  'plot_10c_caption': 'Figure 10c: Substitution frequencies in the quantification window',\n",
       "  'plot_10c_data': [['Nucleotide frequencies in quantification window',\n",
       "    'FANC.Quantification_window_substitution_frequency_table.txt']],\n",
       "  'plot_9_roots': ['9.FANC.Alleles_frequency_table_around_hi',\n",
       "   '9.FANC.Alleles_frequency_table_around_dear',\n",
       "   '9.FANC.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_9_captions': ['Figure 9: Visualization of the distribution of identified alleles around the cleavage site for the hi (GGAATCCCTTCTGCAGCACC). Nucleotides are indicated by unique colors (A = green; C = red; G = yellow; T = purple). Substitutions are shown in bold font. Red rectangles highlight inserted sequences. Horizontal dashed lines indicate deleted sequences. The vertical dashed line indicates the predicted cleavage site.',\n",
       "   'Figure 9: Visualization of the distribution of identified alleles around the cleavage site for the dear (AGCCTTGCAGTGGGCGCGCTA). Nucleotides are indicated by unique colors (A = green; C = red; G = yellow; T = purple). Substitutions are shown in bold font. Red rectangles highlight inserted sequences. Horizontal dashed lines indicate deleted sequences. The vertical dashed line indicates the predicted cleavage site.',\n",
       "   'Figure 9: Visualization of the distribution of identified alleles around the cleavage site for the sgRNA CCCACTGAAGGCCC. Nucleotides are indicated by unique colors (A = green; C = red; G = yellow; T = purple). Substitutions are shown in bold font. Red rectangles highlight inserted sequences. Horizontal dashed lines indicate deleted sequences. The vertical dashed line indicates the predicted cleavage site.'],\n",
       "  'plot_9_datas': [[['Allele frequency table',\n",
       "     'FANC.Alleles_frequency_table_around_hi.txt']],\n",
       "   [['Allele frequency table',\n",
       "     'FANC.Alleles_frequency_table_around_dear.txt']],\n",
       "   [['Allele frequency table',\n",
       "     'FANC.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt']]],\n",
       "  'allele_frequency_files': ['FANC.Alleles_frequency_table_around_hi.txt',\n",
       "   'FANC.Alleles_frequency_table_around_dear.txt',\n",
       "   'FANC.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt'],\n",
       "  'plot_10d_roots': ['10d.FANC.Log2_nucleotide_frequency_around_hi',\n",
       "   '10d.FANC.Log2_nucleotide_frequency_around_dear',\n",
       "   '10d.FANC.Log2_nucleotide_frequency_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_10d_captions': ['Figure 10d: Log2 nucleotide frequencies for each position in the plotting window around the hi (GGAATCCCTTCTGCAGCACC). The quantification window is outlined by the dotted box.',\n",
       "   'Figure 10d: Log2 nucleotide frequencies for each position in the plotting window around the dear (AGCCTTGCAGTGGGCGCGCTA). The quantification window is outlined by the dotted box.',\n",
       "   'Figure 10d: Log2 nucleotide frequencies for each position in the plotting window around the sgRNA CCCACTGAAGGCCC. The quantification window is outlined by the dotted box.'],\n",
       "  'plot_10d_datas': [[], [], []],\n",
       "  'plot_10e_roots': ['10e.FANC.Selected_conversion_at_Cs_around_hi',\n",
       "   '10e.FANC.Selected_conversion_at_Cs_around_dear',\n",
       "   '10e.FANC.Selected_conversion_at_Cs_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_10e_captions': ['Figure 10e: Proportion of each base at each nucleotide targeted by base editors in the plotting window around the hi (GGAATCCCTTCTGCAGCACC). The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10e: Proportion of each base at each nucleotide targeted by base editors in the plotting window around the dear (AGCCTTGCAGTGGGCGCGCTA). The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10e: Proportion of each base at each nucleotide targeted by base editors in the plotting window around the sgRNA CCCACTGAAGGCCC. The number of each target base is annotated on the reference sequence at the bottom of the plot.'],\n",
       "  'plot_10e_datas': [[['Nucleotide frequencies at Cs',\n",
       "     'FANC.Selected_nucleotide_frequency_table_around_hi.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'FANC.Selected_nucleotide_frequency_table_around_dear.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'FANC.Selected_nucleotide_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt']]],\n",
       "  'plot_10f_roots': ['10f.FANC.Selected_conversion_no_ref_at_Cs_around_hi',\n",
       "   '10f.FANC.Selected_conversion_no_ref_at_Cs_around_dear',\n",
       "   '10f.FANC.Selected_conversion_no_ref_at_Cs_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_10f_captions': ['Figure 10f: Non-reference base proportions. For target nucleotides in the plotting window, this plot shows the proportion of non-reference (non-C) bases as a percentage of all non-reference sequences. The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10f: Non-reference base proportions. For target nucleotides in the plotting window, this plot shows the proportion of non-reference (non-C) bases as a percentage of all non-reference sequences. The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10f: Non-reference base proportions. For target nucleotides in the plotting window, this plot shows the proportion of non-reference (non-C) bases as a percentage of all non-reference sequences. The number of each target base is annotated on the reference sequence at the bottom of the plot.'],\n",
       "  'plot_10f_datas': [[['Nucleotide frequencies at Cs',\n",
       "     'FANC.Selected_nucleotide_frequency_table_around_hi.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'FANC.Selected_nucleotide_frequency_table_around_dear.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'FANC.Selected_nucleotide_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt']]],\n",
       "  'plot_10g_roots': ['10g.FANC.Selected_conversion_no_ref_scaled_at_Cs_around_hi',\n",
       "   '10g.FANC.Selected_conversion_no_ref_scaled_at_Cs_around_dear',\n",
       "   '10g.FANC.Selected_conversion_no_ref_scaled_at_Cs_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_10g_captions': ['Figure 10g: Non-reference base counts. For target nucleotides in the plotting window, this plot shows the number of non-reference (non-C) bases. The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10g: Non-reference base counts. For target nucleotides in the plotting window, this plot shows the number of non-reference (non-C) bases. The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10g: Non-reference base counts. For target nucleotides in the plotting window, this plot shows the number of non-reference (non-C) bases. The number of each target base is annotated on the reference sequence at the bottom of the plot.'],\n",
       "  'plot_10g_datas': [[['Nucleotide frequencies at Cs',\n",
       "     'FANC.Selected_nucleotide_frequency_table_around_hi.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'FANC.Selected_nucleotide_frequency_table_around_dear.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'FANC.Selected_nucleotide_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt']]],\n",
       "  'plot_4f_root': '4f.FANC.Global_mutations_in_HDR_reads_with_reference_to_FANC',\n",
       "  'plot_4f_caption': 'Figure 4f: Positions of modifications in HDR reads with respect to the reference sequence (FANC). All modifications (including those outside the quantification window) are shown.',\n",
       "  'plot_4f_data': []},\n",
       " 'HDR': {'name': 'HDR',\n",
       "  'sequence': 'CGGCCGGATGTTCCAATCAGTACGCAGAGAGTCGCCGTCTCCAAGGTGAAAGCTGAAGTAGGGCCTTCGCGCACCTCATGGAATCCCTTCTGCAGCTTTTCCGAGCTTCTGGCGGTCTCAAGCACTACCTACGTCAGCACCTGGGACCCCGCCACCGTGCGCCGGGCCTTGCAGTGGGCGCGCTACCTGCGCCACATCCATCGGCGCTTTGGTCGG',\n",
       "  'sequence_length': 216,\n",
       "  'min_aln_score': 80,\n",
       "  'gap_incentive': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),\n",
       "  'sgRNA_cut_points': [95, 181, 165],\n",
       "  'sgRNA_intervals': [[79, 95], [164, 184], [163, 177]],\n",
       "  'sgRNA_sequences': ['GGAATCCCTTCTGCAGCACC',\n",
       "   'GGCCTTGCAGTGGGCGCGCTA',\n",
       "   'CCCACTGCAAGGCCC'],\n",
       "  'sgRNA_plot_cut_points': [False, False, False],\n",
       "  'sgRNA_plot_idxs': [[76,\n",
       "    77,\n",
       "    78,\n",
       "    79,\n",
       "    80,\n",
       "    81,\n",
       "    82,\n",
       "    83,\n",
       "    84,\n",
       "    85,\n",
       "    86,\n",
       "    87,\n",
       "    88,\n",
       "    89,\n",
       "    90,\n",
       "    91,\n",
       "    92,\n",
       "    93,\n",
       "    94,\n",
       "    95,\n",
       "    96,\n",
       "    97,\n",
       "    98,\n",
       "    99,\n",
       "    100,\n",
       "    101,\n",
       "    102,\n",
       "    103,\n",
       "    104,\n",
       "    105,\n",
       "    106,\n",
       "    107,\n",
       "    108,\n",
       "    109,\n",
       "    110,\n",
       "    111,\n",
       "    112,\n",
       "    113,\n",
       "    114,\n",
       "    115],\n",
       "   [162,\n",
       "    163,\n",
       "    164,\n",
       "    165,\n",
       "    166,\n",
       "    167,\n",
       "    168,\n",
       "    169,\n",
       "    170,\n",
       "    171,\n",
       "    172,\n",
       "    173,\n",
       "    174,\n",
       "    175,\n",
       "    176,\n",
       "    177,\n",
       "    178,\n",
       "    179,\n",
       "    180,\n",
       "    181,\n",
       "    182,\n",
       "    183,\n",
       "    184,\n",
       "    185,\n",
       "    186,\n",
       "    187,\n",
       "    188,\n",
       "    189,\n",
       "    190,\n",
       "    191,\n",
       "    192,\n",
       "    193,\n",
       "    194,\n",
       "    195,\n",
       "    196,\n",
       "    197,\n",
       "    198,\n",
       "    199,\n",
       "    200,\n",
       "    201],\n",
       "   [146,\n",
       "    147,\n",
       "    148,\n",
       "    149,\n",
       "    150,\n",
       "    151,\n",
       "    152,\n",
       "    153,\n",
       "    154,\n",
       "    155,\n",
       "    156,\n",
       "    157,\n",
       "    158,\n",
       "    159,\n",
       "    160,\n",
       "    161,\n",
       "    162,\n",
       "    163,\n",
       "    164,\n",
       "    165,\n",
       "    166,\n",
       "    167,\n",
       "    168,\n",
       "    169,\n",
       "    170,\n",
       "    171,\n",
       "    172,\n",
       "    173,\n",
       "    174,\n",
       "    175,\n",
       "    176,\n",
       "    177,\n",
       "    178,\n",
       "    179,\n",
       "    180,\n",
       "    181,\n",
       "    182,\n",
       "    183,\n",
       "    184,\n",
       "    185]],\n",
       "  'sgRNA_names': ['hi', 'dear', ''],\n",
       "  'sgRNA_mismatches': [[], [0], [7]],\n",
       "  'sgRNA_orig_sequences': ['GGAATCCCTTCTGCAGCACC',\n",
       "   'AGCCTTGCAGTGGGCGCGCTA',\n",
       "   'CCCACTGAAGGCCC'],\n",
       "  'sgRNA_include_idxs': [[181, 182], [165, 166]],\n",
       "  'contains_guide': True,\n",
       "  'contains_coding_seq': True,\n",
       "  'exon_positions': [60,\n",
       "   61,\n",
       "   62,\n",
       "   63,\n",
       "   64,\n",
       "   65,\n",
       "   66,\n",
       "   67,\n",
       "   68,\n",
       "   69,\n",
       "   70,\n",
       "   71,\n",
       "   72,\n",
       "   73,\n",
       "   74,\n",
       "   75,\n",
       "   76,\n",
       "   77,\n",
       "   78,\n",
       "   79,\n",
       "   80,\n",
       "   81,\n",
       "   82,\n",
       "   83,\n",
       "   84,\n",
       "   85,\n",
       "   86,\n",
       "   87,\n",
       "   88,\n",
       "   89,\n",
       "   90,\n",
       "   91,\n",
       "   92,\n",
       "   93,\n",
       "   94,\n",
       "   95,\n",
       "   96,\n",
       "   97,\n",
       "   98,\n",
       "   99],\n",
       "  'exon_len_mods': [-11],\n",
       "  'exon_intervals': [[60, 100]],\n",
       "  'splicing_positions': {58, 59, 100, 101},\n",
       "  'include_idxs': [24,\n",
       "   25,\n",
       "   26,\n",
       "   27,\n",
       "   28,\n",
       "   29,\n",
       "   30,\n",
       "   31,\n",
       "   32,\n",
       "   33,\n",
       "   34,\n",
       "   49,\n",
       "   50,\n",
       "   51,\n",
       "   52,\n",
       "   53,\n",
       "   54],\n",
       "  'exclude_idxs': [0,\n",
       "   1,\n",
       "   2,\n",
       "   3,\n",
       "   4,\n",
       "   5,\n",
       "   6,\n",
       "   7,\n",
       "   8,\n",
       "   9,\n",
       "   10,\n",
       "   11,\n",
       "   12,\n",
       "   13,\n",
       "   14,\n",
       "   201,\n",
       "   202,\n",
       "   203,\n",
       "   204,\n",
       "   205,\n",
       "   206,\n",
       "   207,\n",
       "   208,\n",
       "   209,\n",
       "   210,\n",
       "   211,\n",
       "   212,\n",
       "   213,\n",
       "   214,\n",
       "   215],\n",
       "  'idx_cloned_from': 'FANC',\n",
       "  'fw_seeds': ['ATCAGTACGC',\n",
       "   'TACGCAGAGA',\n",
       "   'AGAGAGTCGC',\n",
       "   'GTCGCCGTCT',\n",
       "   'CGTCTCCAAG',\n",
       "   'CCAAGGTGAA',\n",
       "   'GTGAAAGCTG',\n",
       "   'AGCTGAAGTA',\n",
       "   'AAGTAGGGCC',\n",
       "   'GGGCCTTCGC',\n",
       "   'TTCGCGCACC',\n",
       "   'GCACCTCATG',\n",
       "   'TCATGGAATC',\n",
       "   'GAATCCCTTC',\n",
       "   'CCTTCTGCAG',\n",
       "   'TGCAGCTTTT',\n",
       "   'CTTTTCCGAG',\n",
       "   'CCGAGCTTCT',\n",
       "   'CTTCTGGCGG',\n",
       "   'GGCGGTCTCA',\n",
       "   'TCTCAAGCAC',\n",
       "   'AGCACTACCT',\n",
       "   'TACCTACGTC',\n",
       "   'ACGTCAGCAC',\n",
       "   'AGCACCTGGG',\n",
       "   'CTGGGACCCC',\n",
       "   'ACCCCGCCAC',\n",
       "   'GCCACCGTGC',\n",
       "   'CGTGCGCCGG',\n",
       "   'GCCGGGCCTT',\n",
       "   'GCCTTGCAGT',\n",
       "   'GCAGTGGGCG',\n",
       "   'GGGCGCGCTA',\n",
       "   'CGCTACCTGC',\n",
       "   'CCTGCGCCAC',\n",
       "   'GCCACATCCA'],\n",
       "  'rc_seeds': ['GCGTACTGAT',\n",
       "   'TCTCTGCGTA',\n",
       "   'GCGACTCTCT',\n",
       "   'AGACGGCGAC',\n",
       "   'CTTGGAGACG',\n",
       "   'TTCACCTTGG',\n",
       "   'CAGCTTTCAC',\n",
       "   'TACTTCAGCT',\n",
       "   'GGCCCTACTT',\n",
       "   'GCGAAGGCCC',\n",
       "   'GGTGCGCGAA',\n",
       "   'CATGAGGTGC',\n",
       "   'GATTCCATGA',\n",
       "   'GAAGGGATTC',\n",
       "   'CTGCAGAAGG',\n",
       "   'AAAAGCTGCA',\n",
       "   'CTCGGAAAAG',\n",
       "   'AGAAGCTCGG',\n",
       "   'CCGCCAGAAG',\n",
       "   'TGAGACCGCC',\n",
       "   'GTGCTTGAGA',\n",
       "   'AGGTAGTGCT',\n",
       "   'GACGTAGGTA',\n",
       "   'GTGCTGACGT',\n",
       "   'CCCAGGTGCT',\n",
       "   'GGGGTCCCAG',\n",
       "   'GTGGCGGGGT',\n",
       "   'GCACGGTGGC',\n",
       "   'CCGGCGCACG',\n",
       "   'AAGGCCCGGC',\n",
       "   'ACTGCAAGGC',\n",
       "   'CGCCCACTGC',\n",
       "   'TAGCGCGCCC',\n",
       "   'GCAGGTAGCG',\n",
       "   'GTGGCGCAGG',\n",
       "   'TGGATGTGGC'],\n",
       "  'aln_genome': None,\n",
       "  'aln_chr': None,\n",
       "  'aln_start': None,\n",
       "  'aln_end': None,\n",
       "  'aln_strand': None,\n",
       "  'min_cut': 95,\n",
       "  'max_cut': 181,\n",
       "  'y_values_mut': array([ 1, 14,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0]),\n",
       "  'x_bins_mut': array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),\n",
       "  'y_values_ins': array([16,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0]),\n",
       "  'x_bins_ins': array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),\n",
       "  'y_values_del': array([16,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0]),\n",
       "  'x_bins_del': array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15]),\n",
       "  'hdensity': array([ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 16,  0,\n",
       "          0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0]),\n",
       "  'hlengths': array([-15, -14, -13, -12, -11, -10,  -9,  -8,  -7,  -6,  -5,  -4,  -3,\n",
       "          -2,  -1,   0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,\n",
       "          11,  12,  13,  14,  15]),\n",
       "  'center_index': 15,\n",
       "  'ref_plot_name': 'HDR.',\n",
       "  'insertion_pct_vector_filename': 'HDR.Effect_vector_insertion.txt',\n",
       "  'deletion_pct_vector_filename': 'HDR.Effect_vector_deletion.txt',\n",
       "  'substitution_pct_vector_filename': 'HDR.Effect_vector_substitution.txt',\n",
       "  'combined_pct_vector_filename': 'HDR.Effect_vector_combined.txt',\n",
       "  'quant_window_mod_count_filename': 'HDR.Quantification_window_modification_count_vectors.txt',\n",
       "  'mod_count_filename': 'HDR.Modification_count_vectors.txt',\n",
       "  'mod_count_filename_caption': 'A tab-separated file showing the number of modifications for each position in the amplicon. The first row shows the amplicon sequence, and successive rows show the number of reads with insertions (row 2), insertions_left (row 3), deletions (row 4), substitutions (row 5) and the sum of all modifications (row 6).Additionally, the last row shows the number of reads aligned. If an insertion occurs between bases 5 and 6, the insertions vector will be incremented at bases 5 and 6. However, the insertions_left vector will only be incremented at base 5 so the sum of the insertions_left row represents an accurate count of the number of insertions, whereas the sum of the insertions row will yield twice the number of insertions.',\n",
       "  'frameshift_analysis_filename': 'HDR.Frameshift_analysis.txt',\n",
       "  'frameshift_analysis_filename_caption': 'A text file describing the number of noncoding, in-frame, and frameshift mutations. This report file is produced when the amplicon contains a coding sequence.',\n",
       "  'splice_sites_analysis_filename': 'HDR.Splice_sites_analysis.txt',\n",
       "  'splice_sites_analysis_filename_caption': 'A text file describing the number of splicing sites that are unmodified and modified. This file report is produced when the amplicon contains a coding sequence.',\n",
       "  'insertion_pct_vector_noncoding_filename': 'HDR.Effect_vector_insertion_noncoding.txt',\n",
       "  'insertion_pct_vector_noncoding_filename_caption': 'A tab-separated text file with a one-row header that shows the percentage of reads with a noncoding insertion at each base in the HDR sequence. The first column shows the 1-based position of the amplicon, and the second column shows the percentage of reads with a noncoding insertion at that location. This report file is produced when the amplicon contains a coding sequence.',\n",
       "  'deletion_pct_vector_noncoding_filename': 'HDR.Effect_vector_deletion_noncoding.txt',\n",
       "  'deletion_pct_vector_noncoding_filename_caption': 'A tab-separated text file with a one-row header that shows the percentage of reads with a noncoding deletion at each base in the HDR sequence. The first column shows the 1-based position of the amplicon, and the second column shows the percentage of reads with a noncoding deletion at that location. This report file is produced when the amplicon contains a coding sequence.',\n",
       "  'substitution_pct_vector_noncoding_filename': 'HDR.Effect_vector_substitution_noncoding.txt',\n",
       "  'substitution_pct_vector_noncoding_filename_caption': 'A tab-separated text file with a one-row header that shows the percentage of reads with a noncoding substitution at each base in the HDR sequence. The first column shows the 1-based position of the amplicon, and the second column shows the percentage of reads with a nondcoding substitution at that location. This report file is produced when the amplicon contains a coding sequence.',\n",
       "  'indel_histogram_filename': 'HDR.Indel_histogram.txt',\n",
       "  'indel_histogram_filename_caption': 'A tab-separated text file that shows a histogram of the length of indels (both insertions and deletions) in the HDR sequence in the quantification window. Indels outside of the quantification window are not included. The indel_size column shows the number of substitutions, and the fq column shows the number of reads having an indel of that length.',\n",
       "  'insertion_histogram_filename': 'HDR.Insertion_histogram.txt',\n",
       "  'insertion_histogram_filename_caption': 'A tab-separated text file that shows a histogram of the number of insertions in the HDR sequence in the quantification window. Insertions outside of the quantification window are not included. The ins_size column shows the number of insertions, and the fq column shows the number of reads having that number of insertions.',\n",
       "  'deletion_histogram_filename': 'HDR.Deletion_histogram.txt',\n",
       "  'deletion_histogram_filename_caption': 'A tab-separated text file that shows a histogram of the number of deletions in the HDR sequence in the quantification window. Deletions outside of the quantification window are not included. The del_size column shows the number of deletions, and the fq column shows the number of reads having that number of deletions.',\n",
       "  'substitution_histogram_filename': 'HDR.Substitution_histogram.txt',\n",
       "  'substitution_histogram_filename_caption': 'A tab-separated text file that shows a histogram of the number of substitutions in the HDR sequence in the quantification window. Substitutions outside of the quantification window are not included. The sub_size column shows the number of substitutions, and the fq column shows the number of reads having that number of substitutions.',\n",
       "  'quant_window_nuc_freq_filename': 'HDR.Quantification_window_nucleotide_frequency_table.txt',\n",
       "  'quant_window_nuc_pct_filename': 'HDR.Quantification_window_nucleotide_percentage_table.txt',\n",
       "  'nuc_freq_filename': 'HDR.Nucleotide_frequency_table.txt',\n",
       "  'nuc_pct_filename': 'HDR.Nucleotide_percentage_table.txt',\n",
       "  'quant_window_sub_freq_filename': 'HDR.Quantification_window_substitution_frequency_table.txt',\n",
       "  'sub_freq_table_filename': 'HDR.Substitution_frequency_table.txt',\n",
       "  'plot_2a_root': '2a.HDR.Nucleotide_percentage_quilt',\n",
       "  'plot_2a_caption': 'Figure 2a: Nucleotide distribution across amplicon. At each base in the reference amplicon, the percentage of each base as observed in sequencing reads is shown (A = green; C = orange; G = yellow; T = purple). Black bars show the percentage of reads for which that base was deleted. Brown bars between bases show the percentage of reads having an insertion at that position.',\n",
       "  'plot_2a_data': [['Nucleotide frequency table',\n",
       "    'HDR.Nucleotide_frequency_table.txt']],\n",
       "  'plot_2b_roots': ['2b.HDR.Nucleotide_percentage_quilt_around_hi',\n",
       "   '2b.HDR.Nucleotide_percentage_quilt_around_dear',\n",
       "   '2b.HDR.Nucleotide_percentage_quilt_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_2b_captions': ['Figure 2b: Nucleotide distribution around the hi (GGAATCCCTTCTGCAGCACC).',\n",
       "   'Figure 2b: Nucleotide distribution around the dear (AGCCTTGCAGTGGGCGCGCTA).',\n",
       "   'Figure 2b: Nucleotide distribution around the sgRNA CCCACTGAAGGCCC.'],\n",
       "  'plot_2b_datas': [[['Nucleotide frequency in quantification window',\n",
       "     'HDR.Quantification_window_nucleotide_frequency_table.txt']],\n",
       "   [['Nucleotide frequency in quantification window',\n",
       "     'HDR.Quantification_window_nucleotide_frequency_table.txt']],\n",
       "   [['Nucleotide frequency in quantification window',\n",
       "     'HDR.Quantification_window_nucleotide_frequency_table.txt']]],\n",
       "  'plot_3a_root': '3a.HDR.Indel_size_distribution',\n",
       "  'plot_3a_caption': 'Figure 3a: Frequency distribution of alleles with indels (blue) and without indels (red).',\n",
       "  'plot_3a_data': [['Indel histogram', 'HDR.Indel_histogram.txt']],\n",
       "  'plot_3b_root': '3b.HDR.Insertion_deletion_substitutions_size_hist',\n",
       "  'plot_3b_caption': 'Figure 3b: Left panel, frequency distribution of sequence modifications that increase read length with respect to the reference amplicon, classified as insertions (positive indel size). Middle panel, frequency distribution of sequence modifications that reduce read length with respect to the reference amplicon, classified as deletions (negative indel size). Right panel, frequency distribution of sequence modifications that do not alter read length with respect to the reference amplicon, which are classified as substitutions (number of substituted positions shown).',\n",
       "  'plot_3b_data': [['Insertions frequency', 'HDR.Insertion_histogram.txt'],\n",
       "   ['Deletions Frequency', 'HDR.Deletion_histogram.txt'],\n",
       "   ['Substitutions Frequency', 'HDR.Substitution_histogram.txt']],\n",
       "  'plot_4a_root': '4a.HDR.Combined_insertion_deletion_substitution_locations',\n",
       "  'plot_4a_caption': 'Figure 4a: Combined frequency of any modification across the amplicon. Modifications outside of the quantification window are also shown.',\n",
       "  'plot_4a_data': [],\n",
       "  'plot_4b_root': '4b.HDR.Insertion_deletion_substitution_locations',\n",
       "  'plot_4b_caption': 'Figure 4b: Frequency of insertions, deletions, and substitutions across the entire amplicon, including modifications outside of the quantification window.',\n",
       "  'plot_4b_data': [['Modification frequency',\n",
       "    'HDR.Modification_count_vectors.txt']],\n",
       "  'plot_4c_root': '4c.HDR.Quantification_window_insertion_deletion_substitution_locations',\n",
       "  'plot_4c_caption': 'Figure 4c: Frequency of insertions, deletions, and substitutions across the entire amplicon, considering only modifications that overlap with the quantification window.',\n",
       "  'plot_4c_data': [['Modification frequency in quantification window',\n",
       "    'HDR.Quantification_window_modification_count_vectors.txt']],\n",
       "  'plot_4d_root': '4d.HDR.Position_dependent_average_indel_size',\n",
       "  'plot_4d_caption': 'Figure 4d: Position dependent insertion size(left) and deletion size (right), including only modifications that overlap with the quantification window.',\n",
       "  'plot_4d_data': [],\n",
       "  'plot_5_root': '5.HDR.Frameshift_in-frame_mutations_pie_chart',\n",
       "  'plot_5_caption': 'Figure 5: Frameshift analysis of coding sequence reads affected by modifications (unmodified reads are excluded from this analysis).',\n",
       "  'plot_5_data': [],\n",
       "  'plot_6_root': '6.HDR.Frameshift_in-frame_mutation_profiles',\n",
       "  'plot_6_caption': 'Figure 6: Frameshift and in-frame mutagenesis profiles indicating position affected by modification. The y axis shows the number of reads and percentage of all reads in that category (frameshifted (top) or in-frame (bottom)). 0 reads with no length modifications are not shown.',\n",
       "  'plot_6_data': [['Indel histogram for HDR', 'HDR.Indel_histogram.txt']],\n",
       "  'plot_7_root': '7.HDR.Insertion_deletion_substitution_locations_noncoding',\n",
       "  'plot_7_caption': 'Figure 7: Reads with insertions, deletions, and substitutions mapped to reference amplicon position exclusively in noncoding region/s (that is, without mutations affecting coding sequences). The predicted cleavage site is indicated by a vertical dashed line. Only sequence positions directly adjacent to insertions or directly affected by deletions or substitutions are plotted.',\n",
       "  'plot_7_data': [],\n",
       "  'plot_8_root': '8.HDR.Potential_splice_sites_pie_chart',\n",
       "  'plot_8_caption': 'Figure 8: Predicted impact on splice sites. Potential splice sites modified refers to reads in which the either of the two intronic positions adjacent to exon junctions are disrupted.',\n",
       "  'plot_8_data': [],\n",
       "  'plot_10a_root': '10a.HDR.Substitution_frequencies_at_each_bp',\n",
       "  'plot_10a_caption': 'Figure 10a: Substitution frequencies across the amplicon.',\n",
       "  'plot_10a_data': [['Nucleotide frequencies',\n",
       "    'HDR.Nucleotide_frequency_table.txt']],\n",
       "  'plot_10b_root': '10b.HDR.Substitution_frequency_barplot',\n",
       "  'plot_10b_caption': 'Figure 10b: Substitution frequencies across the amplicon.',\n",
       "  'plot_10b_data': [['Nucleotide frequencies',\n",
       "    'HDR.Nucleotide_frequency_table.txt']],\n",
       "  'plot_10c_root': '10c.HDR.Substitution_frequency_barplot_in_quantification_window',\n",
       "  'plot_10c_caption': 'Figure 10c: Substitution frequencies in the quantification window',\n",
       "  'plot_10c_data': [['Nucleotide frequencies in quantification window',\n",
       "    'HDR.Quantification_window_substitution_frequency_table.txt']],\n",
       "  'plot_9_roots': ['9.HDR.Alleles_frequency_table_around_hi',\n",
       "   '9.HDR.Alleles_frequency_table_around_dear',\n",
       "   '9.HDR.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_9_captions': ['Figure 9: Visualization of the distribution of identified alleles around the cleavage site for the hi (GGAATCCCTTCTGCAGCACC). Nucleotides are indicated by unique colors (A = green; C = red; G = yellow; T = purple). Substitutions are shown in bold font. Red rectangles highlight inserted sequences. Horizontal dashed lines indicate deleted sequences. The vertical dashed line indicates the predicted cleavage site.',\n",
       "   'Figure 9: Visualization of the distribution of identified alleles around the cleavage site for the dear (AGCCTTGCAGTGGGCGCGCTA). Nucleotides are indicated by unique colors (A = green; C = red; G = yellow; T = purple). Substitutions are shown in bold font. Red rectangles highlight inserted sequences. Horizontal dashed lines indicate deleted sequences. The vertical dashed line indicates the predicted cleavage site.',\n",
       "   'Figure 9: Visualization of the distribution of identified alleles around the cleavage site for the sgRNA CCCACTGAAGGCCC. Nucleotides are indicated by unique colors (A = green; C = red; G = yellow; T = purple). Substitutions are shown in bold font. Red rectangles highlight inserted sequences. Horizontal dashed lines indicate deleted sequences. The vertical dashed line indicates the predicted cleavage site.'],\n",
       "  'plot_9_datas': [[['Allele frequency table',\n",
       "     'HDR.Alleles_frequency_table_around_hi.txt']],\n",
       "   [['Allele frequency table', 'HDR.Alleles_frequency_table_around_dear.txt']],\n",
       "   [['Allele frequency table',\n",
       "     'HDR.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt']]],\n",
       "  'allele_frequency_files': ['HDR.Alleles_frequency_table_around_hi.txt',\n",
       "   'HDR.Alleles_frequency_table_around_dear.txt',\n",
       "   'HDR.Alleles_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt'],\n",
       "  'plot_10d_roots': ['10d.HDR.Log2_nucleotide_frequency_around_hi',\n",
       "   '10d.HDR.Log2_nucleotide_frequency_around_dear',\n",
       "   '10d.HDR.Log2_nucleotide_frequency_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_10d_captions': ['Figure 10d: Log2 nucleotide frequencies for each position in the plotting window around the hi (GGAATCCCTTCTGCAGCACC). The quantification window is outlined by the dotted box.',\n",
       "   'Figure 10d: Log2 nucleotide frequencies for each position in the plotting window around the dear (AGCCTTGCAGTGGGCGCGCTA). The quantification window is outlined by the dotted box.',\n",
       "   'Figure 10d: Log2 nucleotide frequencies for each position in the plotting window around the sgRNA CCCACTGAAGGCCC. The quantification window is outlined by the dotted box.'],\n",
       "  'plot_10d_datas': [[], [], []],\n",
       "  'plot_10e_roots': ['10e.HDR.Selected_conversion_at_Cs_around_hi',\n",
       "   '10e.HDR.Selected_conversion_at_Cs_around_dear',\n",
       "   '10e.HDR.Selected_conversion_at_Cs_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_10e_captions': ['Figure 10e: Proportion of each base at each nucleotide targeted by base editors in the plotting window around the hi (GGAATCCCTTCTGCAGCACC). The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10e: Proportion of each base at each nucleotide targeted by base editors in the plotting window around the dear (AGCCTTGCAGTGGGCGCGCTA). The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10e: Proportion of each base at each nucleotide targeted by base editors in the plotting window around the sgRNA CCCACTGAAGGCCC. The number of each target base is annotated on the reference sequence at the bottom of the plot.'],\n",
       "  'plot_10e_datas': [[['Nucleotide frequencies at Cs',\n",
       "     'HDR.Selected_nucleotide_frequency_table_around_hi.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'HDR.Selected_nucleotide_frequency_table_around_dear.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'HDR.Selected_nucleotide_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt']]],\n",
       "  'plot_10f_roots': ['10f.HDR.Selected_conversion_no_ref_at_Cs_around_hi',\n",
       "   '10f.HDR.Selected_conversion_no_ref_at_Cs_around_dear',\n",
       "   '10f.HDR.Selected_conversion_no_ref_at_Cs_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_10f_captions': ['Figure 10f: Non-reference base proportions. For target nucleotides in the plotting window, this plot shows the proportion of non-reference (non-C) bases as a percentage of all non-reference sequences. The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10f: Non-reference base proportions. For target nucleotides in the plotting window, this plot shows the proportion of non-reference (non-C) bases as a percentage of all non-reference sequences. The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10f: Non-reference base proportions. For target nucleotides in the plotting window, this plot shows the proportion of non-reference (non-C) bases as a percentage of all non-reference sequences. The number of each target base is annotated on the reference sequence at the bottom of the plot.'],\n",
       "  'plot_10f_datas': [[['Nucleotide frequencies at Cs',\n",
       "     'HDR.Selected_nucleotide_frequency_table_around_hi.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'HDR.Selected_nucleotide_frequency_table_around_dear.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'HDR.Selected_nucleotide_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt']]],\n",
       "  'plot_10g_roots': ['10g.HDR.Selected_conversion_no_ref_scaled_at_Cs_around_hi',\n",
       "   '10g.HDR.Selected_conversion_no_ref_scaled_at_Cs_around_dear',\n",
       "   '10g.HDR.Selected_conversion_no_ref_scaled_at_Cs_around_sgRNA_CCCACTGAAGGCCC'],\n",
       "  'plot_10g_captions': ['Figure 10g: Non-reference base counts. For target nucleotides in the plotting window, this plot shows the number of non-reference (non-C) bases. The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10g: Non-reference base counts. For target nucleotides in the plotting window, this plot shows the number of non-reference (non-C) bases. The number of each target base is annotated on the reference sequence at the bottom of the plot.',\n",
       "   'Figure 10g: Non-reference base counts. For target nucleotides in the plotting window, this plot shows the number of non-reference (non-C) bases. The number of each target base is annotated on the reference sequence at the bottom of the plot.'],\n",
       "  'plot_10g_datas': [[['Nucleotide frequencies at Cs',\n",
       "     'HDR.Selected_nucleotide_frequency_table_around_hi.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'HDR.Selected_nucleotide_frequency_table_around_dear.txt']],\n",
       "   [['Nucleotide frequencies at Cs',\n",
       "     'HDR.Selected_nucleotide_frequency_table_around_sgRNA_CCCACTGAAGGCCC.txt']]]}}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d45ae8b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_total = crispresso2_info['results']['alignment_stats']['counts_total']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfc5da62",
   "metadata": {},
   "source": [
    "We're interested in figures 9 for plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08ba822e",
   "metadata": {},
   "source": [
    "from CRISPRessoCORE\n",
    "```\n",
    "plot_9_input = {\n",
    "    'reference_seq': ref_seq_around_cut,\n",
    "    'df_alleles': df_to_plot,\n",
    "    'fig_filename_root': fig_filename_root,\n",
    "    'custom_colors': custom_config[\"colors\"],\n",
    "    'MIN_FREQUENCY': args.min_frequency_alleles_around_cut_to_plot,\n",
    "    'MAX_N_ROWS': args.max_rows_alleles_around_cut_to_plot,\n",
    "    'SAVE_ALSO_PNG': save_png,\n",
    "    'plot_cut_point': plot_cut_point,\n",
    "    'sgRNA_intervals': new_sgRNA_intervals,\n",
    "    'sgRNA_names': sgRNA_names,\n",
    "    'sgRNA_mismatches': sgRNA_mismatches,\n",
    "    'annotate_wildtype_allele': args.annotate_wildtype_allele,\n",
    "}\n",
    "debug('Plotting allele distribution around cut for {0}'.format(ref_name))\n",
    "plot(CRISPRessoPlot.plot_alleles_table, plot_9_input)\n",
    "```\n",
    "\n",
    "`df_to_plot` is constructed with the following lines:\n",
    "```\n",
    "df_alleles_around_cut=CRISPRessoShared.get_dataframe_around_cut(df_alleles.loc[df_alleles['Reference_Name'] == ref_name], cut_point, plot_half_window)\n",
    "count_total = counts_total[ref_name]\n",
    "if args.allele_plot_pcts_only_for_assigned_reference:\n",
    "    df_alleles_around_cut['%AllReads']=df_alleles_around_cut['%Reads']\n",
    "    df_alleles_around_cut['%Reads']=df_alleles_around_cut['#Reads']/count_total*100\n",
    "\n",
    "#...\n",
    "\n",
    "df_to_plot = df_alleles_around_cut\n",
    "if not args.expand_allele_plots_by_quantification:\n",
    "    df_to_plot = df_alleles_around_cut.groupby(['Aligned_Sequence', 'Reference_Sequence']).sum().reset_index().set_index('Aligned_Sequence')\n",
    "    df_to_plot.sort_values(by=['#Reads', 'Aligned_Sequence', 'Reference_Sequence'], inplace=True, ascending=[False, True, True])\n",
    "```\n",
    "\n",
    "Need to figure out what the cut point is supposed to be?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f3c6377d",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispresso2_info_copy = crispresso2_info.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a550ade7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sgRNA_sequences = refs[ref_name]['sgRNA_sequences']\n",
    "sgRNA_cut_points = refs[ref_name]['sgRNA_cut_points']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a9154687",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GGAATCCCTTCTGCAGCACC', 'GGCCTTGCAGTGGGCGCGCTA', 'CCCACTGCAAGGCCC']\n",
      "[91, 188, 172]\n"
     ]
    }
   ],
   "source": [
    "print(sgRNA_sequences)\n",
    "print(sgRNA_cut_points)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d09a33e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crispresso2_info['running_info']['args'].plot_window_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "91a16785",
   "metadata": {},
   "outputs": [],
   "source": [
    "cut_point = sgRNA_cut_points[0]\n",
    "plot_half_window = max(1, crispresso2_info['running_info']['args'].plot_window_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b57b78f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "crispresso2_info['running_info']['args'].coding_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f0b3e307",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'ref_positions'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/micromamba/envs/jupyter-env/lib/python3.10/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'ref_positions'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[16], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m df_alleles_around_cut\u001b[38;5;241m=\u001b[39m\u001b[43mCRISPRessoShared\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_dataframe_around_cut\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf_alleles\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mdf_alleles\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mReference_Name\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mref_name\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcut_point\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mplot_half_window\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m df_to_plot \u001b[38;5;241m=\u001b[39m df_alleles_around_cut\n",
      "File \u001b[0;32m~/Edilytics/CRISPResso2/CRISPResso2/CRISPRessoShared.py:1213\u001b[0m, in \u001b[0;36mget_dataframe_around_cut\u001b[0;34m(df_alleles, cut_point, offset, collapse_by_sequence)\u001b[0m\n\u001b[1;32m   1207\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (cut_point \u001b[38;5;241m+\u001b[39m offset \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;241m>\u001b[39m \u001b[38;5;28mlen\u001b[39m(ref1)):\n\u001b[1;32m   1208\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m (BadParameterException(\n\u001b[1;32m   1209\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mThe plotting window cannot extend past the end of the amplicon. Amplicon length is \u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(\n\u001b[1;32m   1210\u001b[0m             \u001b[38;5;28mlen\u001b[39m(ref1)) \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m but plot extends to \u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(cut_point \u001b[38;5;241m+\u001b[39m offset \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m)))\n\u001b[1;32m   1212\u001b[0m df_alleles_around_cut \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\n\u001b[0;32m-> 1213\u001b[0m     \u001b[38;5;28mlist\u001b[39m(\u001b[43mdf_alleles\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mlambda\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mrow\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mget_row_around_cut\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrow\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcut_point\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moffset\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mvalues),\n\u001b[1;32m   1214\u001b[0m     columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAligned_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReference_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUnedited\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_deleted\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_inserted\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_mutated\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#Reads\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m   1215\u001b[0m              \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mReads\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m   1217\u001b[0m df_alleles_around_cut \u001b[38;5;241m=\u001b[39m df_alleles_around_cut\u001b[38;5;241m.\u001b[39mgroupby(\n\u001b[1;32m   1218\u001b[0m     [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAligned_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReference_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUnedited\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_deleted\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_inserted\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m   1219\u001b[0m      \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_mutated\u001b[39m\u001b[38;5;124m'\u001b[39m])\u001b[38;5;241m.\u001b[39msum()\u001b[38;5;241m.\u001b[39mreset_index()\u001b[38;5;241m.\u001b[39mset_index(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAligned_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m   1221\u001b[0m df_alleles_around_cut\u001b[38;5;241m.\u001b[39msort_values(by\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#Reads\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAligned_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReference_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m], inplace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, ascending\u001b[38;5;241m=\u001b[39m[\u001b[38;5;28;01mFalse\u001b[39;00m, \u001b[38;5;28;01mTrue\u001b[39;00m, \u001b[38;5;28;01mTrue\u001b[39;00m])\n",
      "File \u001b[0;32m~/micromamba/envs/jupyter-env/lib/python3.10/site-packages/pandas/core/frame.py:10374\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[0;34m(self, func, axis, raw, result_type, args, by_row, engine, engine_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m  10360\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcore\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mapply\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m frame_apply\n\u001b[1;32m  10362\u001b[0m op \u001b[38;5;241m=\u001b[39m frame_apply(\n\u001b[1;32m  10363\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m  10364\u001b[0m     func\u001b[38;5;241m=\u001b[39mfunc,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  10372\u001b[0m     kwargs\u001b[38;5;241m=\u001b[39mkwargs,\n\u001b[1;32m  10373\u001b[0m )\n\u001b[0;32m> 10374\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapply\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/micromamba/envs/jupyter-env/lib/python3.10/site-packages/pandas/core/apply.py:916\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    913\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw:\n\u001b[1;32m    914\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_raw(engine\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine, engine_kwargs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine_kwargs)\n\u001b[0;32m--> 916\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/micromamba/envs/jupyter-env/lib/python3.10/site-packages/pandas/core/apply.py:1063\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1061\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply_standard\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m   1062\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mengine \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpython\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m-> 1063\u001b[0m         results, res_index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_series_generator\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1064\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1065\u001b[0m         results, res_index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_series_numba()\n",
      "File \u001b[0;32m~/micromamba/envs/jupyter-env/lib/python3.10/site-packages/pandas/core/apply.py:1081\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1078\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m option_context(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmode.chained_assignment\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m   1079\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i, v \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(series_gen):\n\u001b[1;32m   1080\u001b[0m         \u001b[38;5;66;03m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[0;32m-> 1081\u001b[0m         results[i] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mv\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1082\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[1;32m   1083\u001b[0m             \u001b[38;5;66;03m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[1;32m   1084\u001b[0m             \u001b[38;5;66;03m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[1;32m   1085\u001b[0m             results[i] \u001b[38;5;241m=\u001b[39m results[i]\u001b[38;5;241m.\u001b[39mcopy(deep\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "File \u001b[0;32m~/Edilytics/CRISPResso2/CRISPResso2/CRISPRessoShared.py:1213\u001b[0m, in \u001b[0;36mget_dataframe_around_cut.<locals>.<lambda>\u001b[0;34m(row)\u001b[0m\n\u001b[1;32m   1207\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (cut_point \u001b[38;5;241m+\u001b[39m offset \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;241m>\u001b[39m \u001b[38;5;28mlen\u001b[39m(ref1)):\n\u001b[1;32m   1208\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m (BadParameterException(\n\u001b[1;32m   1209\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mThe plotting window cannot extend past the end of the amplicon. Amplicon length is \u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(\n\u001b[1;32m   1210\u001b[0m             \u001b[38;5;28mlen\u001b[39m(ref1)) \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m but plot extends to \u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(cut_point \u001b[38;5;241m+\u001b[39m offset \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m)))\n\u001b[1;32m   1212\u001b[0m df_alleles_around_cut \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\n\u001b[0;32m-> 1213\u001b[0m     \u001b[38;5;28mlist\u001b[39m(df_alleles\u001b[38;5;241m.\u001b[39mapply(\u001b[38;5;28;01mlambda\u001b[39;00m row: \u001b[43mget_row_around_cut\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrow\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcut_point\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moffset\u001b[49m\u001b[43m)\u001b[49m, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\u001b[38;5;241m.\u001b[39mvalues),\n\u001b[1;32m   1214\u001b[0m     columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAligned_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReference_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUnedited\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_deleted\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_inserted\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_mutated\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#Reads\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m   1215\u001b[0m              \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mReads\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m   1217\u001b[0m df_alleles_around_cut \u001b[38;5;241m=\u001b[39m df_alleles_around_cut\u001b[38;5;241m.\u001b[39mgroupby(\n\u001b[1;32m   1218\u001b[0m     [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAligned_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReference_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUnedited\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_deleted\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_inserted\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m   1219\u001b[0m      \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_mutated\u001b[39m\u001b[38;5;124m'\u001b[39m])\u001b[38;5;241m.\u001b[39msum()\u001b[38;5;241m.\u001b[39mreset_index()\u001b[38;5;241m.\u001b[39mset_index(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAligned_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m   1221\u001b[0m df_alleles_around_cut\u001b[38;5;241m.\u001b[39msort_values(by\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#Reads\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAligned_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReference_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m], inplace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, ascending\u001b[38;5;241m=\u001b[39m[\u001b[38;5;28;01mFalse\u001b[39;00m, \u001b[38;5;28;01mTrue\u001b[39;00m, \u001b[38;5;28;01mTrue\u001b[39;00m])\n",
      "File \u001b[0;32m~/Edilytics/CRISPResso2/CRISPResso2/CRISPRessoShared.py:1195\u001b[0m, in \u001b[0;36mget_row_around_cut\u001b[0;34m(row, cut_point, offset)\u001b[0m\n\u001b[1;32m   1194\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_row_around_cut\u001b[39m(row, cut_point, offset):\n\u001b[0;32m-> 1195\u001b[0m     cut_idx \u001b[38;5;241m=\u001b[39m \u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mref_positions\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mindex(cut_point)\n\u001b[1;32m   1196\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAligned_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m][cut_idx \u001b[38;5;241m-\u001b[39m offset \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m:cut_idx \u001b[38;5;241m+\u001b[39m offset \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m], row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mReference_Sequence\u001b[39m\u001b[38;5;124m'\u001b[39m][\n\u001b[1;32m   1197\u001b[0m                                                                                cut_idx \u001b[38;5;241m-\u001b[39m offset \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m:cut_idx \u001b[38;5;241m+\u001b[39m offset \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m], \\\n\u001b[1;32m   1198\u001b[0m            row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRead_Status\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUNMODIFIED\u001b[39m\u001b[38;5;124m'\u001b[39m, row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_deleted\u001b[39m\u001b[38;5;124m'\u001b[39m], row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_inserted\u001b[39m\u001b[38;5;124m'\u001b[39m], row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_mutated\u001b[39m\u001b[38;5;124m'\u001b[39m], row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#Reads\u001b[39m\u001b[38;5;124m'\u001b[39m], \\\n\u001b[1;32m   1199\u001b[0m            row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124mReads\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[0;32m~/micromamba/envs/jupyter-env/lib/python3.10/site-packages/pandas/core/series.py:1121\u001b[0m, in \u001b[0;36mSeries.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1118\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[key]\n\u001b[1;32m   1120\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m key_is_scalar:\n\u001b[0;32m-> 1121\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1123\u001b[0m \u001b[38;5;66;03m# Convert generator to list before going through hashable part\u001b[39;00m\n\u001b[1;32m   1124\u001b[0m \u001b[38;5;66;03m# (We will iterate through the generator there to check for slices)\u001b[39;00m\n\u001b[1;32m   1125\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n",
      "File \u001b[0;32m~/micromamba/envs/jupyter-env/lib/python3.10/site-packages/pandas/core/series.py:1237\u001b[0m, in \u001b[0;36mSeries._get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m   1234\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[label]\n\u001b[1;32m   1236\u001b[0m \u001b[38;5;66;03m# Similar to Index.get_value, but we do not fall back to positional\u001b[39;00m\n\u001b[0;32m-> 1237\u001b[0m loc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1239\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(loc):\n\u001b[1;32m   1240\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[loc]\n",
      "File \u001b[0;32m~/micromamba/envs/jupyter-env/lib/python3.10/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'ref_positions'"
     ]
    }
   ],
   "source": [
    "df_alleles_around_cut=CRISPRessoShared.get_dataframe_around_cut(df_alleles.loc[df_alleles['Reference_Name'] == ref_name], cut_point, plot_half_window)\n",
    "df_to_plot = df_alleles_around_cut\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef1d29c9",
   "metadata": {},
   "source": [
    "## Amino Acid coding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "3f924fba",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_seq = crispresso2_info['running_info']['args'].coding_seq\n",
    "# coding_seq = coding_seq[1::]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "118287cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_to_amino_acid = {\n",
    "    'TTT': 'Phe', 'TTC': 'Phe', 'TTA': 'Leu', 'TTG': 'Leu',\n",
    "    'TCT': 'Ser', 'TCC': 'Ser', 'TCA': 'Ser', 'TCG': 'Ser',\n",
    "    'TAT': 'Tyr', 'TAC': 'Tyr', 'TAA': 'STOP', 'TAG': 'STOP',\n",
    "    'TGT': 'Cys', 'TGC': 'Cys', 'TGA': 'STOP', 'TGG': 'Trp',\n",
    "    'CTT': 'Leu', 'CTC': 'Leu', 'CTA': 'Leu', 'CTG': 'Leu',\n",
    "    'CCT': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',\n",
    "    'CAT': 'His', 'CAC': 'His', 'CAA': 'Gln', 'CAG': 'Gln',\n",
    "    'CGT': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg',\n",
    "    'ATT': 'Ile', 'ATC': 'Ile', 'ATA': 'Ile', 'ATG': 'Met',\n",
    "    'ACT': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',\n",
    "    'AAT': 'Asn', 'AAC': 'Asn', 'AAA': 'Lys', 'AAG': 'Lys',\n",
    "    'AGT': 'Ser', 'AGC': 'Ser', 'AGA': 'Arg', 'AGG': 'Arg',\n",
    "    'GTT': 'Val', 'GTC': 'Val', 'GTA': 'Val', 'GTG': 'Val',\n",
    "    'GCT': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',\n",
    "    'GAT': 'Asp', 'GAC': 'Asp', 'GAA': 'Glu', 'GAG': 'Glu',\n",
    "    'GGT': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "3ab9b3a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGG -> Gly\n",
      "CCT -> Pro\n",
      "TCG -> Ser\n",
      "CGC -> Arg\n",
      "ACC -> Thr\n",
      "TCA -> Ser\n",
      "TGG -> Trp\n",
      "AAT -> Asn\n",
      "CCC -> Pro\n",
      "TTC -> Phe\n",
      "TGC -> Cys\n",
      "AGC -> Ser\n",
      "ACC -> Thr\n",
      "TGG -> Trp\n",
      "ATC -> Ile\n",
      "GCT -> Ala\n",
      "TTT -> Phe\n"
     ]
    }
   ],
   "source": [
    "while len(coding_seq) > 2:\n",
    "    codon, coding_seq = coding_seq[:3], coding_seq[3:]\n",
    "    print(f\"{codon} -> {codon_to_amino_acid[codon]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "97fde43f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_to_plot = pd.read_csv('df_to_plot.csv', header=0, sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "12ea4a35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reference_Sequence</th>\n",
       "      <th>Unedited</th>\n",
       "      <th>n_deleted</th>\n",
       "      <th>n_inserted</th>\n",
       "      <th>n_mutated</th>\n",
       "      <th>#Reads</th>\n",
       "      <th>%Reads</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aligned_Sequence</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG</th>\n",
       "      <td>CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>163</td>\n",
       "      <td>78.365385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CATGGAATCCCTTCTGCA---CCTGGATCGCTTTTCCGAG</th>\n",
       "      <td>CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>3.846154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CATGGAATCCCTTCTG----ACCTGGATCGCTTTTCCGAG</th>\n",
       "      <td>CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.961538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CATGGAATCCCTTCTGCAGC--CTGGATCGCTTTTCCGAG</th>\n",
       "      <td>CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.961538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CATGGAATCCCTTCTGCAGCCACCTGGATCGCTTTTCCGA</th>\n",
       "      <td>CATGGAATCCCTTCTGCAGC-ACCTGGATCGCTTTTCCGA</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.961538</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                Reference_Sequence  \\\n",
       "Aligned_Sequence                                                                     \n",
       "CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG  CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG   \n",
       "CATGGAATCCCTTCTGCA---CCTGGATCGCTTTTCCGAG  CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG   \n",
       "CATGGAATCCCTTCTG----ACCTGGATCGCTTTTCCGAG  CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG   \n",
       "CATGGAATCCCTTCTGCAGC--CTGGATCGCTTTTCCGAG  CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG   \n",
       "CATGGAATCCCTTCTGCAGCCACCTGGATCGCTTTTCCGA  CATGGAATCCCTTCTGCAGC-ACCTGGATCGCTTTTCCGA   \n",
       "\n",
       "                                          Unedited  n_deleted  n_inserted  \\\n",
       "Aligned_Sequence                                                            \n",
       "CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG         1          0           0   \n",
       "CATGGAATCCCTTCTGCA---CCTGGATCGCTTTTCCGAG         1          0           0   \n",
       "CATGGAATCCCTTCTG----ACCTGGATCGCTTTTCCGAG         1          0           0   \n",
       "CATGGAATCCCTTCTGCAGC--CTGGATCGCTTTTCCGAG         1          0           0   \n",
       "CATGGAATCCCTTCTGCAGCCACCTGGATCGCTTTTCCGA         1          0           0   \n",
       "\n",
       "                                          n_mutated  #Reads     %Reads  \n",
       "Aligned_Sequence                                                        \n",
       "CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG          0     163  78.365385  \n",
       "CATGGAATCCCTTCTGCA---CCTGGATCGCTTTTCCGAG          0       8   3.846154  \n",
       "CATGGAATCCCTTCTG----ACCTGGATCGCTTTTCCGAG          0       2   0.961538  \n",
       "CATGGAATCCCTTCTGCAGC--CTGGATCGCTTTTCCGAG          0       2   0.961538  \n",
       "CATGGAATCCCTTCTGCAGCCACCTGGATCGCTTTTCCGA          0       2   0.961538  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_to_plot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6690ab9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_9a_inputs = {\n",
    "    'reference_seq': 'CATGGAATCCCTTCTGCAGCACCTGGATCGCTTTTCCGAG', \n",
    "    'df_alleles': df_to_plot, \n",
    "    'fig_filename_root': './scripts/figures/9a', \n",
    "    'custom_colors': {\n",
    "        'Substitution': '#0000FF', \n",
    "        'Insertion': '#008000', \n",
    "        'Deletion': '#FF0000', \n",
    "        'A': '#7FC97F', \n",
    "        'T': '#BEAED4', \n",
    "        'C': '#FDC086', \n",
    "        'G': '#FFFF99', \n",
    "        'N': '#C8C8C8', \n",
    "        '-': '#1E1E1E'\n",
    "    }, \n",
    "    'MIN_FREQUENCY': 0.2, \n",
    "    'MAX_N_ROWS': 50, \n",
    "    'SAVE_ALSO_PNG': True, \n",
    "    'plot_cut_point': False, \n",
    "    'sgRNA_intervals': [(3, 22), (99, 119), (98, 112)], \n",
    "    'sgRNA_names': ['hi', 'dear', ''], \n",
    "    'sgRNA_mismatches': [[], [0], [7]], \n",
    "    'annotate_wildtype_allele': ''\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "86094be1",
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    X, # 2d array: This is the sequence converted to ints I THINK for the cmap \n",
    "    annot, # This is the sequence of bp's \n",
    "    y_labels, # this is the percentage and tallies displayed to the right ``\n",
    "    insertion_dict, # I think this is key: which aligned_seq has insertion, value: where the insertion is\n",
    "    per_element_annot_kws, # dict of dicts: this is for bolding the substitutions\n",
    "    is_reference) = CRISPRessoPlot.prep_amino_acid_table(\n",
    "        df_to_plot, \n",
    "        plot_9a_inputs['reference_seq'], \n",
    "        plot_9a_inputs['MAX_N_ROWS'], \n",
    "        plot_9a_inputs['MIN_FREQUENCY']\n",
    "        )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "58e63f86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff4ec5f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO:\n",
    "# 1. change X to be the sequence of amino acids converted to ints for coloring\n",
    "# 2. change annot to be the sequence of amino acids\n",
    "# 3. change or get rid of color map. If its easy then let's just make it all the same color\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "891f12df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(per_element_annot_kws[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "78099829",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CRISPRessoPlot.plot_amino_acid_table(**plot_9a_inputs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02840d80",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
